File: libs/server/api/src/api/model/node_model.rs
use serde::Serialize;
use std::collections::BTreeMap;

use crate::domain::{DiagramNode, JsonObject, Position, Ref};

use super::super::links::{
    workspace_diagram_href, workspace_diagram_node_href, workspace_diagram_nodes_href,
    workspace_logical_entity_href, Link,
};

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(in crate::api) struct NodeModel {
    #[serde(rename = "_links")]
    links: BTreeMap<String, Link>,
    id: String,
    kind: String,
    parent: Option<Ref<String>>,
    position: Position,
    width: Option<i64>,
    height: Option<i64>,
    data: JsonObject,
    created_at: String,
    updated_at: String,
}

pub(in crate::api) fn node_model(workspace_id: &str, node: &DiagramNode) -> NodeModel {
    let diagram_id = node.diagram_id();
    let node_id = node.identity();
    let description = node.description();

    let mut links = BTreeMap::from([
        (
            "self".to_string(),
            Link::new(workspace_diagram_node_href(
                workspace_id,
                diagram_id,
                node_id,
            )),
        ),
        (
            "collection".to_string(),
            Link::new(workspace_diagram_nodes_href(workspace_id, diagram_id)),
        ),
        (
            "diagram".to_string(),
            Link::new(workspace_diagram_href(workspace_id, diagram_id)),
        ),
    ]);

    if let Some(logical_entity) = &description.logical_entity {
        links.insert(
            "logicalEntity".to_string(),
            Link::new(workspace_logical_entity_href(
                workspace_id,
                logical_entity.id(),
            )),
        );
    }

    NodeModel {
        links,
        id: node_id.to_string(),
        kind: description.kind.clone(),
        parent: description.parent.clone(),
        position: description.position.clone(),
        width: description.width,
        height: description.height,
        data: description.data.clone(),
        created_at: node.created_at().to_string(),
        updated_at: node.updated_at().to_string(),
    }
}

File: apps/server-nest/src/api/model/node-model.ts
import { DiagramNode, LogicalEntity, Ref } from '../../domain';
import {
  link,
  Link,
  workspaceDiagramHref,
  workspaceDiagramNodeHref,
  workspaceDiagramNodesHref,
  workspaceLogicalEntityHref,
} from '../links';
import { logicalEntityModel, LogicalEntityModel } from './logical-entity-model';

interface NodeEmbedded {
  'logical-entity': LogicalEntityModel;
}

export interface NodeModel {
  _links: Record<string, Link>;
  _embedded?: NodeEmbedded;
  id: string;
  kind: string;
  parent: Ref<string> | null;
  position: ReturnType<DiagramNode['description']>['position'];
  width: number | null;
  height: number | null;
  data: ReturnType<DiagramNode['description']>['data'];
  createdAt: string;
  updatedAt: string;
}

export function nodeModel(
  workspaceId: string,
  node: DiagramNode,
  logicalEntity: LogicalEntity | null,
): NodeModel {
  const nodeId = node.identity();
  const description = node.description();
  const diagramId = description.diagram.id();
  const links: Record<string, Link> = {
    self: link(workspaceDiagramNodeHref(workspaceId, diagramId, nodeId)),
    collection: link(workspaceDiagramNodesHref(workspaceId, diagramId)),
    diagram: link(workspaceDiagramHref(workspaceId, diagramId)),
  };
  if (description.logicalEntity) {
    links['logical-entity'] = link(
      workspaceLogicalEntityHref(workspaceId, description.logicalEntity.id()),
    );
  }

  const model: NodeModel = {
    _links: links,
    id: nodeId,
    kind: description.kind,
    parent: description.parent,
    position: description.position,
    width: description.width,
    height: description.height,
    data: description.data,
    createdAt: description.createdAt,
    updatedAt: description.updatedAt,
  };
  if (logicalEntity) {
    model._embedded = { 'logical-entity': logicalEntityModel(logicalEntity) };
  }
  return model;
}
